File: 프로그래머스/lv4/49995. 쿠키 구입/쿠키 구입.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<int> cookie){
    int answer = 0;

    for(int i=0; i<cookie.size() - 1; i++){
        int suml = cookie[i];
        int sumr = cookie[i + 1];
        int l_idx = i;
        int r_idx = i + 1;

        while(1){

            if(suml == sumr)
                answer = max(answer, suml);

            if(suml > sumr){
                if(r_idx + 1 == cookie.size()) break;
                sumr += cookie[++r_idx];
            }
            else{
                if(l_idx - 1 < 0) break;
                suml += cookie[--l_idx];
            }            

        }
    }

    return answer;
}

File: 프로그래머스/4/49995. 쿠키 구입/쿠키 구입.java
class Solution {
    private static int[] sumCookie;
    public int solution(int[] cookie) {
        int answer = 0;
        sumCookie = new int[cookie.length + 1];

        for (int i = 1; i < sumCookie.length; i++) {
            sumCookie[i] = sumCookie[i - 1] + cookie[i - 1];
        }

        for (int l = 0; l < cookie.length - 2; l++) {
            for (int r = cookie.length-1; r > l; r--) {
                answer = Math.max(answer, cal(l, r));
            }
        }

        return answer;
    }

    private static int cal(int l, int r) {
        int son1, son2, mid;
        int low = l, high = r;

        while (high > low) {
            mid = (high + low) / 2;
            son1 = sumCookie[mid + 1] - sumCookie[l];
            son2 = sumCookie[r + 1] - sumCookie[mid + 1];

            if (son1 == son2) {
                return son1;
            } else if (son1 > son2) {
                high = mid;
            } else {
                low = mid+1;
            }
        }
        return 0;
    }
}